README.md
# Flutter - Drawer Component

A Flutter based Reusable Drawer component which provides Drawer menu with configurable items and dynamic views that can be used in any Flutter application.


## Table of contents
- [Flutter Support](#Flutter-support)
- [Demo](#demo)
- [Features](#features)
- [Getting started](#getting-started)
- [Usage](#usage)
- [Methods](#methods)
- [Want to Contribute?](#want-to-contribute)
- [Need Help / Support?](#need-help)
- [Collection of Components](#collection-of-Components)
- [Changelog](#changelog)
- [License](#license)
- [Keywords](#Keywords)


## Flutter Support

Version - Flutter 1.17 (stable)

We have tested our program in above version, however you can use it in other versions as well.


## Demo
[![](drawer.gif)](https://github.com/weblineindia/Flutter-Drawer/blob/master/drawer.gif)

------

## Features

* Drawer with Customizable Menu header.
* Drawer with Customizable Menu icons.
* Customizable UI with trailing items.


## Getting started

* Download this sample project and import app_drawer.dart Widget file in your Flutter App. 
* Update Header, Title based on your requirements. 


## Usage

Setup process is described below to integrate in project.

## Configure DrawerMenu in your widget by following below method steps

### Methods

Step 1.Add  'provider' package to pubspec.yaml

Step 2. Add MultiProvider in Your application widget like this
       
    // Put DrawerNavigationProvider as mention below
    class MyApp extends StatelessWidget {
      @override
      Widget build(BuildContext context) {
        final MaterialColor colorCustom = MaterialColor(0xFFFACA58, color);
        return MultiProvider(
          providers: [
            ChangeNotifierProvider.value(
              value: DrawerNavigationProvider(),
            )
          ],
          child: MaterialApp(
            title: 'Flutter Demo',
            home: ...,
          ),
        );
      }
    }
       

Step 3. User following DrawerVC constructor for adding drawer to user main screen             
       
       Color selectedTextColor;
       Color textColor;
       int selectedIndex;
       bool isHeader = true;
       bool isShowIcon = true;
       Widget headerWidget;
       List<DrawerMenuItem> arrDrawerItems;
       Function(int) didItemSelected;
       
       DrawerVC({
            this.arrDrawerItems,
            this.isHeader,
            this.headerWidget,
            this.isShowIcon,
            this.textColor,
            this.selectedTextColor,
            this.didItemSelected,
        });
        
        // Pass DrawerMenuItem Array to the DrawerVC's Constructor
        class DrawerMenuItem {
          String title;
          IconData icon;
          Widget subItemWidget;

          // ignore: sort_constructors_first
          DrawerMenuItem({this.title, this.icon, this.subItemWidget});
        }


Step 4. Pass header widget if you want to add it otherwise skip it  

    this.isHeader = true // if you want to add header make sure you need to set this flag to true
    headerWidget: Widget() 


Step 5.if you want set text color  and selected text color 

    textColor: Colors.white
    selectedTextColor: Colors.white
    
Step 6.if you want enable/disable Drawer item icons 
    
    isShowIcon : true // if you want to show icon the set this to true otherwise false

Step 7.if you want add trailing item icons then pass widget to  DrawerMenuItem class using below constructor   
        
      subItemWidget : Widget() // use this parameter to set trailing icons to drawer
       
    // Pass DrawerMenuItem Array to the DrawerVC's Constructor
           class DrawerMenuItem {
             String title;
             IconData icon;
             Widget subItemWidget;

             // ignore: sort_constructors_first
             DrawerMenuItem({this.title, this.icon, this.subItemWidget});
           }
Step 8. if you want selected index of drawer menu you can get that in following ways 

    class HomeScreen extends StatelessWidget {
      final arrDrawerItems = [
        DrawerMenuItem(title: 'Home', icon: Icons.home, subItemWidget: null),
        DrawerMenuItem(
          title: 'Second View',
          icon: Icons.favorite,
          subItemWidget: const Text('13'),
        ),
        DrawerMenuItem(
            title: 'Third View', icon: Icons.search, subItemWidget: null),
      ]; //List for drawer items

      @override
      Widget build(BuildContext context) {
        return Scaffold(
          drawer: DrawerVC(
            arrDrawerItems: arrDrawerItems,
            headerWidget: _buildHeaderForDrawer(context),
            isHeader: true,
            isShowIcon: true,
            textColor: Colors.black,
            selectedTextColor: Colors.amber,
            didItemSelected: (int selectedIndex) {} // Using this function you get selected index of drawer,
          ),
          appBar: AppBar(
              centerTitle: true,
              iconTheme: const IconThemeData(color: Colors.white),
              // You must implement this Consumer<DrawerNavigationProvider> provide for navigation and  you can also get selected index of drawer
              title: Consumer<DrawerNavigationProvider>(
                builder: (ctx, item, child) {
                  if (item.selectedIndex == 0) {
                    return const Text(
                      'Home',
                      style: TextStyle(color: Colors.white),
                    );
                  } else if (item.selectedIndex == 1) {
                    return const Text('Second View',
                        style: TextStyle(color: Colors.white));
                  } else {
                    return const Text('Third View',
                        style: TextStyle(color: Colors.white));
                  }
                },
              ),
              brightness: Brightness.light),
          body: Center(
          // You must implement this Consumer<DrawerNavigationProvider> provide for navigation and  you can also get selected index of drawer
            child: Consumer<DrawerNavigationProvider>(
                  builder: (ctx, item, child) {
                    if (item.selectedIndex == 0) {
                      return YOUR_WIDGET();
                    } else if (item.selectedIndex == 1) {
                      return YOUR_WIDGET();
                    } else {
                      return YOUR_WIDGET();
                    }
                  },
                ) ??
                const Text('No data'),
          ),
        );
      }
    }
    
------

## Want to Contribute?

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).

------

## Need Help? 
If you need help then feel free to contact our [Flutter developers](https://www.weblineindia.com/flutter-cross-platform-mobile-app-development.html).

 ------
 
## Collection of Components
 We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development.](https://www.weblineindia.com/software-development-resources.html)
 
------

## Changelog
Detailed changes for each release are documented in [CHANGELOG](./CHANGELOG).

## License

[MIT](LICENSE)

[mit]: https://github.com/weblineindia/Flutter-Drawer/blob/master/LICENSE

## Keywords
Flutter Drawer Menu, Custom Drawer menu, Side menu, FLutter Slider Menu, Custom Slider, Weblineindia


CHANGELOG
1.0.0 (July 17, 2020)
Initial release - Sample source code for Flutter Customizable Drawer Menu.


lib/main.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'app_drawer.dart';
import 'home_Screen.dart';

/*
Author Name  :  WeblineIndia  |  https://www.weblineindia.com/

For more such software development components and code libraries, visit us at
https://www.weblineindia.com/software-development-resources.html 

Our Github URL : https://github.com/weblineindia
*/

void main() => runApp(MyApp());

// ignore: must_be_immutable
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  Map<int, Color> color = {
    50: const Color.fromRGBO(250, 202, 88, .1),
    100: const Color.fromRGBO(250, 202, 88, .2),
    200: const Color.fromRGBO(250, 202, 88, .3),
    300: const Color.fromRGBO(250, 202, 88, .4),
    400: const Color.fromRGBO(250, 202, 88, .5),
    500: const Color.fromRGBO(250, 202, 88, .6),
    600: const Color.fromRGBO(250, 202, 88, .7),
    700: const Color.fromRGBO(250, 202, 88, .8),
    800: const Color.fromRGBO(250, 202, 88, .9),
    900: const Color.fromRGBO(250, 202, 88, 1),
  };

  @override
  Widget build(BuildContext context) {
    final MaterialColor colorCustom = MaterialColor(0xFFFACA58, color);
    return MultiProvider(
      providers: [
        ChangeNotifierProvider.value(
          value: DrawerNavigationProvider(),
        )
      ],
      child: MaterialApp(
        title: 'Flutter Demo',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: colorCustom,
          accentColor: Colors.white,
        ),
        home: HomeScreen(),
      ),
    );
  }
}


lib/app_drawer.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

// ignore: must_be_immutable
class DrawerVC extends StatelessWidget {
  Color selectedTextColor;
  Color textColor;
  int selectedIndex;
  bool isHeader = true;
  bool isShowIcon = true;
  Widget headerWidget;
  List<DrawerMenuItem> arrDrawerItems;
  Function(int) didItemSelected;

  // ignore: sort_constructors_first
  DrawerVC({
    this.arrDrawerItems,
    this.isHeader,
    this.headerWidget,
    this.isShowIcon,
    this.textColor,
    this.selectedTextColor,
    this.didItemSelected,
  });

  @override
  Widget build(BuildContext context) {
    final DrawerNavigationProvider navigationProvider =
        Provider.of<DrawerNavigationProvider>(context, listen: false);

    return Drawer(
      child: ListView.builder(
        padding: const EdgeInsets.all(0.0),
        itemCount: arrDrawerItems.length + 1,
        itemBuilder: (ctx, index) {
          final int idx = index - 1;
          return index == 0
              ? isHeader
                  ? headerWidget ??
                      SizedBox(
                        height: MediaQuery.of(context).size.height * 0.08,
                      )
                  : SizedBox(
                      height: MediaQuery.of(context).size.height * 0.08,
                    )
              : GestureDetector(
                  child: Padding(
                    padding: const EdgeInsets.all(16.0),
                    child: Row(
                      children: [
                        if (isShowIcon)
                          Container(
                              margin: const EdgeInsets.only(
                                  left: 0, top: 0, right: 10, bottom: 0),
                              child: Icon(arrDrawerItems[idx].icon,
                                  color: idx == selectedIndex
                                      ? selectedTextColor
                                      : textColor))
                        else
                          Container(),
                        Expanded(
                          child: Text(
                            arrDrawerItems[idx].title,
                            style: TextStyle(
                                color: idx == selectedIndex
                                    ? selectedTextColor
                                    : textColor),
                          ),
                        ),
                        Container(
                            margin: const EdgeInsets.only(
                                left: 10, top: 0, right: 0, bottom: 0),
                            child: arrDrawerItems[idx].subItemWidget ??
                                Container()),
                      ],
                    ),
                  ),
                  onTap: () {
                    didItemSelected(idx);
                    selectedIndex = idx;
                    navigationProvider.setDraweNavigation(idx);
                    Navigator.pop(context);
                  },
                );
        },
      ),
    );
  }

//   void navigateToSelectedScreen(int selectedItemIndex, BuildContext context) {
//     final DrawerProvider baseData =
//         Provider.of<DrawerProvider>(context, listen: false);

//     if (baseData.drawerOption == selectedItemIndex) {
//       Navigator.of(context).pop();
//       return;
//     }

//     switch (selectedItemIndex) {
//       case DrawerOption.home:
//         baseData.setBaseData(
//             title: 'Quotes Category', selectedDrawerOption: DrawerOption.home);
//         break;

//       case DrawerOption.categories:
//         baseData.setBaseData(
//             title: 'Favourites', selectedDrawerOption: DrawerOption.categories);
//         break;

//       case DrawerOption.aboutUs:
//         baseData.setBaseData(
//             title: 'Search', selectedDrawerOption: DrawerOption.aboutUs);
//         break;

//       default:
//         break;
//     }
//     Navigator.of(context).pop();
//   }
}

class DrawerMenuItem {
  String title;
  IconData icon;
  Widget subItemWidget;

  // ignore: sort_constructors_first
  DrawerMenuItem({this.title, this.icon, this.subItemWidget});
}

class DrawerNavigationProvider with ChangeNotifier {
  int selectedIndex = 0;

  void setDraweNavigation(int index) {
    selectedIndex = index;
    notifyListeners();
  }
}


lib/home_Screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'app_drawer.dart';
import 'drawer_screens/category.dart';
import 'drawer_screens/home.dart';

class HomeScreen extends StatelessWidget {
  final arrDrawerItems = [
    DrawerMenuItem(title: 'First View', icon: Icons.home, subItemWidget: null),
    DrawerMenuItem(
      title: 'Second View',
      icon: Icons.favorite,
      subItemWidget: const Text('13'),
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      drawer: DrawerVC(
        arrDrawerItems: arrDrawerItems,
        headerWidget: _buildHeaderForDrawer(context),
        isHeader: true,
        isShowIcon: true,
        textColor: Colors.black,
        selectedTextColor: Colors.amber,
        didItemSelected: (int selectedIndex) {},
      ),
      appBar: AppBar(
          centerTitle: true,
          iconTheme: const IconThemeData(color: Colors.white),
          title: Consumer<DrawerNavigationProvider>(
            builder: (ctx, item, child) {
              if (item.selectedIndex == 0) {
                return const Text(
                  'First View',
                  style: TextStyle(color: Colors.white),
                );
              } else if (item.selectedIndex == 1) {
                return const Text('Second View',
                    style: TextStyle(color: Colors.white));
              } else {
                return null;
              }
            },
          ),
          brightness: Brightness.light),
      body: Center(
        child: Consumer<DrawerNavigationProvider>(
              builder: (ctx, item, child) {
                if (item.selectedIndex == 0) {
                  return Home();
                } else if (item.selectedIndex == 1) {
                  return Category();
                } else {
                  return null;
                }
              },
            ) ??
            const Text('No data'),
      ),
    );
  }

  Widget _buildHeaderForDrawer(BuildContext context) {
    return DrawerHeader(
      decoration: BoxDecoration(color: Theme.of(context).primaryColor),
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            Image.asset(
              'assets/header.png',
              height: 72.0,
              width: 72.0,
            ),
            const SizedBox(height: 12),
            const Text(
              'Title',
              style: TextStyle(fontSize: 18, color: Colors.white),
            ),
            const SizedBox(height: 4),
            const Text(
              'SubTitle',
              style: TextStyle(color: Colors.white),
            ),
          ],
        ),
      ),
      margin: const EdgeInsets.all(0),
      padding: const EdgeInsets.all(0),
    );
  }
}
